### PR TITLE
Optimizations: Foreach replaced by indexed loops

### PR DESCRIPTION
**Pull request by feliperussi**
## Description

The changes included in this pull request are suggestions to modify some parts of the code to optimize it. Looking through the source code, we identified some for each method that could be easily changed to indexed loops to ensure optimization for the code and reduce possible security and memory risks. 

## Issues
Fixes #

In lib/demos/material/tabs_demo.dart:

Old method with forEach:
image

New method with indexed loop:
   
```
   Widget build(BuildContext context) {
    final localizations = GalleryLocalizations.of(context)!;
    final tabs = [
      localizations.colorsRed,
      localizations.colorsOrange,
      localizations.colorsGreen,
    ];

    List<Tab> tabWidgets = [];
    List<Text> tabCenters = [];

    for (int i = 0; i < tabs.length; i++) {
      tabWidgets.add(Tab(text: tabs[i]));
      tabCenters.add(Text(tabs[i]));
    }

    return Scaffold(
      appBar: AppBar(
        automaticallyImplyLeading: false,
        title: Text(
          localizations.demoTabsNonScrollingTitle,
        ),
        bottom: TabBar(
          controller: _tabController,
          isScrollable: false,
          tabs: tabWidgets,
        ),
      ),
      body: TabBarView(
        controller: _tabController,
        children: tabCenters,
      ),
    );
  }
```

In lib/routes.dart:

Old method with forEach:
image

New method with indexed loop:
```
static Route<dynamic>? onGenerateRoute(RouteSettings settings) {
  for (int i = 0; i < paths.length; i++) {
    final path = paths[i];
    final regExpPattern = RegExp(path.pattern);
    if (regExpPattern.hasMatch(settings.name!)) {
      final firstMatch = regExpPattern.firstMatch(settings.name!)!;
      final match = (firstMatch.groupCount == 1) ? firstMatch.group(1) : null;
      if (kIsWeb) {
        return NoAnimationMaterialPageRoute<void>(
          builder: (context) => path.builder(context, match),
          settings: settings,
        );
      }
      if (path.openInSecondScreen) {
        return TwoPanePageRoute<void>(
          builder: (context) => path.builder(context, match),
          settings: settings,
        );
      } else {
        return MaterialPageRoute<void>(
          builder: (context) => path.builder(context, match),
          settings: settings,
        );
      }
    }
  }

  // If no match was found, we let [WidgetsApp.onUnknownRoute] handle it.
  return null;
}
```

In lib/studies/shrine/app.dart:

Old method with forEach:
image

New method with indexed loop:

```
AppStateModel fromPrimitives(Object? data) {
  final appState = AppStateModel()..loadProducts();
  final appData = Map<String, dynamic>.from(data as Map);

  // Reset selected category.
  final categoryIndex = appData['category_index'] as int;
  appState.setCategory(categories[categoryIndex]);

  // Reset cart items.
  final cartItems = appData['cart_data'] as Map<dynamic, dynamic>;
  final cartEntries = cartItems.entries.toList();
  for (int i = 0; i < cartEntries.length; i++) {
    dynamic id = cartEntries[i].key;
    dynamic quantity = cartEntries[i].value;
    appState.addMultipleProductsToCart(id as int, quantity as int);
  }

  return appState;
}
```

After suggesting those changes, the app continues to function as before. We hope that they are well received by the community. This app is really useful!
---

>**Note**: Please find the development and releasing instructions at https://github.com/flutter/gallery#development

### CODE REVIEW

**Comment by domesticmouse** [comment]
Please sign the CLA. We can't land the PR without it =)

File: lib/demos/material/tabs_demo.dart [hunk 1]
```diff
@@ -180,19 +188,12 @@ class __TabsNonScrollableDemoState extends State<_TabsNonScrollableDemo>
         bottom: TabBar(
           controller: _tabController,
           isScrollable: false,
-          tabs: [
```
Comment (domesticmouse) [inline lib/demos/material/tabs_demo.dart:183]: Please revert this change.

File: lib/routes.dart [hunk 2]
```diff
@@ -122,7 +122,8 @@ class RouteConfiguration {
     RouteSettings settings,
     bool hasHinge,
   ) {
-    for (final path in paths) {
+    for (int i = 0; i < paths.length; i++) {
```
Comment (domesticmouse) [inline lib/routes.dart:125]: What is the intent of this change? The previous code is clearer in intent, and avoids potential off by one errors.

File: lib/demos/material/tabs_demo.dart [hunk 3]
```diff
@@ -180,19 +188,12 @@ class __TabsNonScrollableDemoState extends State<_TabsNonScrollableDemo>
...
-            for (final tab in tabs) Tab(text: tab),
-          ],
+          tabs: tabWidgets,
         ),
       ),
       body: TabBarView(
         controller: _tabController,
-        children: [
```
Comment (domesticmouse) [inline lib/demos/material/tabs_demo.dart:190]: Please revert this change.

File: lib/studies/shrine/app.dart [hunk 4]
```diff
@@ -196,9 +196,12 @@ class _RestorableAppStateModel extends RestorableListenable<AppStateModel> {
 
     // Reset cart items.
     final cartItems = appData['cart_data'] as Map<dynamic, dynamic>;
-    cartItems.forEach((dynamic id, dynamic quantity) {
+    final cartEntries = cartItems.entries.toList();
```
Comment (domesticmouse) [inline lib/studies/shrine/app.dart:199]: Please use `for (final cart in cartItems) { ... }` as this approach is verbose and error prone.

File: lib/studies/shrine/app.dart [hunk 5]
```diff
@@ -196,9 +196,12 @@ class _RestorableAppStateModel extends RestorableListenable<AppStateModel> {
...
+    for (int i = 0; i < cartEntries.length; i++) {
+      dynamic id = cartEntries[i].key;
```
Comment (domesticmouse) [inline lib/studies/shrine/app.dart:201]: I'd be tempted to use Dart 3 destructing here.

**Review by domesticmouse — CHANGES_REQUESTED** [review]

**Comment by Hixie** [comment]
@feliperussi Will you be able to continue working on this? It's totally fine if not; we can take it over (though it might take longer). I'm just asking because I'm trying to clean out our review queue.

**Comment by Hixie** [comment]
I'm going to close this as it looks like it is no longer being worked on.

**Pull request closed without merging**